File: LeetCode/JAVA/622.java
class MyCircularQueue {

    private int[] q;
    private int front;
    private int size;
    private int capacity;

    public MyCircularQueue(int k) {
        q = new int[k];
        front = size = 0;
        capacity = k;
    }
    
    public boolean enQueue(int value) {
        if(isFull()) {
            return false;
        }

        q[(front + size) % capacity] = value;
        size++;
        return true;
    }
    
    public boolean deQueue() {
        if(isEmpty()) {
            return false;
        }

        front = (front + 1) % capacity;
        size--;
        return true;
    }
    
    public int Front() {
        return isEmpty() ? -1 : q[front];
    }
    
    public int Rear() {
        return isEmpty() ? -1 : q[(front + size - 1) % capacity];
    }
    
    public boolean isEmpty() {
        return size == 0;
    }
    
    public boolean isFull() {
        return size == capacity;
    }
}

File: LeetCode/C++/622.cpp
class MyCircularQueue {

private:
    vector<int> q;
    int front;
    int size;

public:
    MyCircularQueue(int k) {
        q.resize(k, 0);
        front = 0;
        size = 0;
    }
    
    bool enQueue(int value) {
        if(isFull()) {
            return false;
        } 

        q[(front + size) % q.size()] = value;
        size++;
        return true;
    }
    
    bool deQueue() {
        if(isEmpty()) {
            return false;
        }

        front = (front + 1) % q.size();
        size--;
        return true;
    }
    
    int Front() {
        return isEmpty() ? -1 : q[front];
    }
    
    int Rear() {
        return isEmpty() ? -1 : q[(front + size - 1) % q.size()];
    }
    
    bool isEmpty() {
        return size == 0;
    }
    
    bool isFull() {
        return q.size() == size;
    }
};
